cmd/basic/main.go

```diff
@@ -78,152 +82,204 @@ func main() {
-// - zipf or uniform random
-// - requests per connect (-1 for unlim)
-// - gets per etc
-// - multiget or not
-// - set or add to replace
-// - delete frequency
-// - set size range
-// - variances: peak/antipeak load
-// - variances: how often to change item sizes
-type BasicLoader struct {
-	servers               []string
-	stopAfter             time.Time
-	desiredConnCount      int
-	requestsPerSleep      int
-	requestBundlesPerConn int
-	sleepPerBundle        time.Duration
-	setValueSizes         []int
-	deletePercent         int
-	keyLength             int
-	keyPrefix             string
-	keySpace              int
-	keyTTL                uint
-	useZipf               bool
-	zipfS                 float64 // (> 1, generally 1.01-2) pulls the power curve toward 0)
-	zipfV                 float64 // v (< keySpace) puts the main part of the curve before this number
-	valueSize             uint
-	clientFlags           uint
+type Config struct {
+	ClientFlags    uint
+	ConnCount      int
+	DelRatio       int
+	Duration       time.Duration
+	GetRatio       int
+	KeyLength      int
+	KeyPrefix      string
+	KeySpace       int
+	KeyTTL         uint
+	Pipelines      uint
+	RngSeed        int64
+	RPS            int
+	Servers        []string
+	SetRatio       int
+	Socket         string
+	StripKeyPrefix bool
+	UseZipf        bool
+	ValueSize      uint
+	WarmPercent    int
+	ZipfS          float64 // (> 1, generally 1.01-2) pulls the power curve toward 0)
+	ZipfV          float64 // v (< keySpace) puts the main part of the curve before this number
+	tachymeter     *tachymeter.Tachymeter
 }
 
-func (l *BasicLoader) Run() {
-	var runners int
-	// TODO: should be method of surfacing errors.
-	doneChan := make(chan int, 50)
+func (conf *Config) Run() (err error) {
+	g, _ := errgroup.WithContext(context.Background())
 
-	for {
-		for runners < l.desiredConnCount {
-			go l.Worker(doneChan)
-			runners++
-		}
-		res := <-doneChan
-		if res == 0 {
-			//fmt.Println("That's a bingo!")
-		}
-		runners--
-		if *cpuprofile != "" && time.Now().After(l.stopAfter) {
+	samples := conf.RPS * conf.ConnCount
+	if samples < 1000 {
+		samples = 1000
+	}
+
+	if conf.WarmPercent > 0 {
+		err = conf.WarmCache()
+		if err != nil {
 			return
 		}
 	}
+
+	threadStats := make(chan Stats, conf.ConnCount)
+	conf.tachymeter = tachymeter.New(&tachymeter.Config{Size: samples})
+	startTime := time.Now()
+
+	for worker := 0; worker < conf.ConnCount; worker++ {
+		index := worker
+		g.Go(func() error {
+			return conf.Worker(index, threadStats)
+		})
+	}
+
+	err = g.Wait()
+	endTime := time.Now()
+	if err != nil {
+		return
+	}
+
+	conf.tachymeter.SetWallTime(time.Since(startTime))
+	close(threadStats)
+	testStats := &Stats{}
+	for stats := range threadStats {
+		testStats.Add(&stats)
+	}
+
+	report := &Report{
+		StartTime: startTime,
+		EndTime:   endTime,
+		Config:    conf,
+		Metrics:   conf.tachymeter.Calc(),
+		Stats:     testStats,
+	}
+	err = report.PrettyPrint()
+
+	return
```

Comment:
Everything else here is just waiting for the goroutines to finish and collecting stats/metrics.